From cc8b4ffe5729281c2f9ac7fbfef631278157736f Mon Sep 17 00:00:00 2001
From: Chris Wilson <chris@chris-wilson.co.uk>
Date: Tue, 22 Nov 2016 14:41:18 +0000
Subject: [PATCH 2001/4301] drm/i915: Don't deref context->file_priv ERR_PTR
 upon reset

commit d9e9da64c4a762493c5ac24b3b2bbc43c2754786 upstream

When a user context is closed, it's file_priv backpointer is replaced by
ERR_PTR(-EBADF); be careful not to chase this invalid pointer after a
hang and a GPU reset.

Signed-off-by: Chris Wilson <chris@chris-wilson.co.uk>
Fixes: b083a0870c79 ("drm/i915: Add per client max context ban limit")
Cc: Mika Kuoppala <mika.kuoppala@intel.com>
Reviewed-by: Mika Kuoppala <mika.kuoppala@intel.com>
Link: http://patchwork.freedesktop.org/patch/msgid/20161122144121.7379-1-chris@chris-wilson.co.uk
---
 drivers/gpu/drm/i915/i915_gem.c |   12 ++++--------
 1 files changed, 4 insertions(+), 8 deletions(-)

diff --git a/drivers/gpu/drm/i915/i915_gem.c b/drivers/gpu/drm/i915/i915_gem.c
index b38d4f8..b0dd0f0 100644
--- a/drivers/gpu/drm/i915/i915_gem.c
+++ b/drivers/gpu/drm/i915/i915_gem.c
@@ -2647,16 +2647,12 @@ static void i915_gem_context_mark_guilty(struct i915_gem_context *ctx)
 			 ctx->name, ctx->ban_score,
 			 yesno(ctx->banned));
 
-	if (!ctx->file_priv)
+	if (!ctx->banned || IS_ERR_OR_NULL(ctx->file_priv))
 		return;
 
-	if (ctx->banned) {
-		ctx->file_priv->context_bans++;
-
-		DRM_DEBUG_DRIVER("client %s has has %d context banned\n",
-				 ctx->name,
-				 ctx->file_priv->context_bans);
-	}
+	ctx->file_priv->context_bans++;
+	DRM_DEBUG_DRIVER("client %s has had %d context banned\n",
+			 ctx->name, ctx->file_priv->context_bans);
 }
 
 static void i915_gem_context_mark_innocent(struct i915_gem_context *ctx)
-- 
1.7.5.4

